src/state.rs

```diff
@@ -174,6 +215,43 @@ impl ProxyState {
         }
         None
     }
+
+    fn set_gateway_address(&self, us: &mut Upstream, hbone_port: u16) -> anyhow::Result<()> {
+        let workload_ip = self.load_balance(&us.workload)?;
+        if us.workload.gateway_address.is_none() {
+            us.workload.gateway_address = Some(match us.workload.protocol {
+                Protocol::HBONE => {
+                    let ip = us
+                        .workload
+                        .waypoint_svc_ip_address()?
+                        .unwrap_or(workload_ip);
+                    SocketAddr::from((ip, hbone_port))
+                }
+                Protocol::TCP => SocketAddr::from((workload_ip, us.port)),
+            });
+        }
+        Ok(())
+    }
+
+    pub fn load_balance(&self, workload: &Workload) -> Result<IpAddr, Error> {
+        // TODO: add more sophisticated routing logic, perhaps based on ipv4/ipv6 support underneath us.
+        // if/when we support that, this function may need to move to get access to the necessary metadata.
+        // Randomly pick an IP
+        // TODO: do this more efficiently, and not just randomly
+        let Some(ip) = workload.workload_ips.choose(&mut rand::thread_rng()) else {
+
+            if !workload.hostname.is_empty() {
+                let ip = self
+                    .resolved_dns
+                    .load_balance_for_workload(workload.clone().uid)?;
```

Comment:
Can you remind me our current status on "dns Services"? is it a TODO, pending, we killed it? I forgot entirely :-)

I absolutely don't mind starting with one, just want to make sure I understand the end state

Reply:
im not sure exactly what you mean? what is a "dns service"?

Reply:
bump on this